Sources/FirebaseCore/FutureProtocol.swift

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: BSD-3-Clause
+
+@_exported
+import firebase
+
+import Foundation
+
+@_spi(Internal)
+public typealias FutureCompletionType = swift_firebase.swift_cxx_shims.firebase.FutureCompletionType
+
+// This protocol enables extracting common code for Future handling. Because
+// C++ interop is limited for templated class types, we need to invent a
+// protocol to reflect the features of a Future<R> that should be generically
+// available. This works by having a C++ annotation (see ConformingFuture<R>)
+// that specifies this protocol conformance.
+@_spi(Internal)
+public protocol FutureProtocol {
+  associatedtype ResultType
+  func error() -> Int32
+  func __error_messageUnsafe() -> UnsafePointer<CChar>?
+  func __resultUnsafe() -> UnsafePointer<ResultType>?
+  func CallOnCompletion(_ completion: FutureCompletionType, _ user_data: UnsafeMutableRawPointer?)
+}
+
+// A home for various helper functions that make it easier to work with
+// Firebase Future objects from Swift.
+@_spi(Internal)
+public extension FutureProtocol {
+  // Callsites retain their own reference to the Future<R>, but they still need
+  // a way to know when the Future completes. This provides that mechanism.
+  // While the underlying Firebase `OnCompletion` method can provide a reference
+  // back to the Future, we don't need to expose that here.
+  func setCompletion(_ completion: @escaping () -> Void) {
+    CallOnCompletion({ ptr in
+      Unmanaged<CompletionWrapper>.fromOpaque(ptr!).takeRetainedValue().completion()
```

Comment:
Just to double check my memory understanding here the `passRetained` effectively emits a +1 ref count on the object, and then `takeRetainedValue` decrements that ref count right so the ptr will get released at the end of this scope?

Reply:
That's the outcome yeah. Can think of `takeRetainedValue` as adopting the pointer and storing it into a managed container that will then release the object when that container goes out of scope.